Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py

```diff
@@ -0,0 +1,308 @@
+# SPDX - License - Identifier: GPL - 3.0 +
+
+from mantid.api import DataProcessorAlgorithm, MultipleFileProperty, FileAction, MatrixWorkspaceProperty, FileProperty
+from mantid.kernel import Direction
+from mantid.simpleapi import *
+
+import numpy as np
+from typing import List
+
+
+class ILLLagrange(DataProcessorAlgorithm):
+
+    progress = None
+    output_ws_name = None
+    empty_cell_ws = None
+    water_correction = None
+    intermediate_workspaces = None
+
+    INCIDENT_ENERGY_OFFSET = 4.5
+
+    use_incident_energy = False
+    convert_to_wavenumber = False
+
+    # max difference between two identical points, two points closer than that will be merged in the merge algorithm
+    EPSILON = 1e-2
+
+    # default header size in data files in case it couldn't be determined
+    HEADER_SIZE_DEFAULT = 36
+
+    # default column indices in data files in case they couldn't be determined
+    ENERGY_COL_DEFAULT = 1
+    MONITOR_COUNT_COL_DEFAULT = 2
+    DETECTOR_COUNT_COL_DEFAULT = 5
+
+    @staticmethod
+    def category():
+        return 'ILL\\Indirect'
+
+    @staticmethod
+    def summary():
+        return 'Reduces Lagrange data.'
+
+    @staticmethod
+    def seeAlso():
+        return []
+
+    @staticmethod
+    def name():
+        return 'ILLLagrange'
+
+    def setup(self):
+        self.use_incident_energy = self.getProperty('UseIncidentEnergy').value
+        self.convert_to_wavenumber = self.getProperty('ConvertToWaveNumber').value
+
+    def PyInit(self):
+        self.declareProperty(MultipleFileProperty('SampleRuns', action=FileAction.Load, extensions=['']),
+                             doc='Sample run(s).')
+        self.declareProperty(MultipleFileProperty('ContainerRuns', action=FileAction.OptionalLoad, extensions=['']),
+                             doc='Container run(s) (empty cell)')
+        self.declareProperty(FileProperty('CorrectionFile', '', action=FileAction.OptionalLoad, extensions=['.txt']),
+                             doc='Correction reference file.')
+        self.declareProperty(MatrixWorkspaceProperty('OutputWorkspace', '', direction=Direction.Output),
+                             doc='The output workspace containing reduced data.')
+        self.declareProperty(name='UseIncidentEnergy', defaultValue=False,
+                             doc='Show the energies as incident energies, not transfer ones.')
+        self.declareProperty(name='ConvertToWaveNumber', defaultValue=False,
+                             doc='Convert axis unit to energy in wave number (cm-1)')
+
+        # the list of all the intermediate workspaces to group at the end
+        self.intermediate_workspaces = []
+
+    def PyExec(self):
+        self.setup()
+
+        correction_file = self.getPropertyValue('CorrectionFile')
+        self.output_ws_name = self.getPropertyValue('OutputWorkspace')
+
+        # load correction
+        if correction_file:
+            self.water_correction = np.loadtxt(correction_file)
```

Comment:
I would also add a try/catch and a check on `self.water_correction` to properly report any failure when reading the file.